File: 298. Binary Tree Longest Consecutive Sequence.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def longestConsecutive(self, root: TreeNode) -> int:
        if not root:
            return 0
        return max(self.dfs(root.left, root.val, 1), self.dfs(root.right, root.val, 1))
        
        
    def dfs(self, node, prev_value, length):
        if not node:
            return length
        if node.val == prev_value+1:
            return max(self.dfs(node.left, node.val, length+1), self.dfs(node.right, node.val, length+1))
        else:
            return max(length, self.dfs(node.left, node.val, 1), self.dfs(node.right, node.val, 1))

File: 298. Binary Tree Longest Consecutive Sequence.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int ans;
    public int longestConsecutive(TreeNode root) {
        ans = 0;
        helper(root, null, 0);
        return ans;
    }
    private void helper(TreeNode root, TreeNode parent, int len) {
        if (root==null) {return;}
        len = (parent!=null && root.val == parent.val+1) ? len+1 : 1;
        ans = Math.max(len, ans);
        helper(root.left, root, len);
        helper(root.right, root, len);
    }
}